action/protocol/staking/handler_candidate_selfstake.go

```diff
@@ -0,0 +1,150 @@
+		cand.AddVote(p.calculateVoteWeight(prevBucket, false))
+		cand.SelfStake = big.NewInt(0)
+	}
+	// change bucket candidate
+	if !address.Equal(bucket.Candidate, cand.Owner) {
+		if err = p.changeBucketCandidate(csm, bucket, bucketCand, cand); err != nil {
+			return log, nil, err
+		}
+	}
+	// bind new bucket
+	cand.SelfStakeBucketIdx = bucket.Index
+	cand.SelfStake = big.NewInt(0).SetBytes(bucket.StakedAmount.Bytes())
+	cand.SubVote(p.calculateVoteWeight(bucket, false))
+	cand.AddVote(p.calculateVoteWeight(bucket, true))
+
+	if err = csm.Upsert(cand); err != nil {
+		return log, nil, csmErrorToHandleError(cand.Owner.String(), err)
+	}
+
+	return log, txLogs, nil
+}
+
+func (p *Protocol) validateBucketSelfStake(ctx context.Context, csm CandidateStateManager, esm *EndorsementStateManager, bucket *VoteBucket, cand *Candidate) ReceiptError {
+	blkCtx := protocol.MustGetBlockCtx(ctx)
+	if err := validateBucketMinAmount(bucket, p.config.RegistrationConsts.MinSelfStake); err != nil {
+		return err
+	}
+	if err := validateBucketStake(bucket, true); err != nil {
+		return err
+	}
+	if err := validateBucketSelfStake(csm, bucket, false); err != nil {
+		return err
+	}
+	if err := validateBucketCandidate(bucket, cand.Owner); err != nil {
+		return err
+	}
+	if validateBucketOwner(bucket, cand.Owner) != nil &&
+		validateBucketEndorsement(esm, bucket, true, blkCtx.BlockHeight) != nil {
+		return &handleError{
+			err:           errors.New("bucket is not a self-owned or endorsed bucket"),
+			failureStatus: iotextypes.ReceiptStatus_ErrUnauthorizedOperator,
+		}
+	}
+	return nil
+}
+
+func (p *Protocol) changeBucketCandidate(csm CandidateStateManager, bucket *VoteBucket, prevCand, cand *Candidate) error {
+	// update bucket index
+	if err := csm.delCandBucketIndex(bucket.Candidate, bucket.Index); err != nil {
+		return err
+	}
+	if err := csm.putCandBucketIndex(cand.Owner, bucket.Index); err != nil {
+		return err
+	}
+	// update bucket candidate
+	bucket.Candidate = cand.Owner
+	if err := csm.updateBucket(bucket.Index, bucket); err != nil {
+		return err
+	}
+	// update previous candidate
+	weightedVotes := p.calculateVoteWeight(bucket, false)
+	if err := prevCand.SubVote(weightedVotes); err != nil {
+		return &handleError{
+			err:           errors.Wrapf(err, "failed to subtract vote for previous candidate %s", prevCand.Owner.String()),
+			failureStatus: iotextypes.ReceiptStatus_ErrNotEnoughBalance,
+		}
+	}
+	if err := csm.Upsert(prevCand); err != nil {
+		return csmErrorToHandleError(prevCand.Owner.String(), err)
+	}
+
+	// update current candidate
+	if err := cand.AddVote(weightedVotes); err != nil {
+		return &handleError{
+			err:           errors.Wrapf(err, "failed to add vote for candidate %s", cand.Owner.String()),
+			failureStatus: iotextypes.ReceiptStatus_ErrInvalidBucketAmount,
+		}
+	}
+	if err := csm.Upsert(cand); err != nil {
+		return csmErrorToHandleError(cand.Owner.String(), err)
+	}
```

Comment:
skip?

Reply:
function deleted, b/c bucket must be voted to the candidate